Use JSON serialization for Job into ORMJob

EliKlein opened the pull request:
## Summary

Serializing Job objects with JSON, in order to change a PickleType column to String.

## References

#6316
Jobs are now unable to accept args/kwargs that need to be kept as a specific type (e.g. if you wanted to pass one job as an argument to another job, you'd use its id, not an object representing it)
Jobs now do not save Exceptions as the same type they were initially raised (they become ErrorSavedWithJob)
Jobs now do not save results

The results might be able to be changed to be saved if needed, but Exception typing would be very difficult if possible, and type'd args/kwargs are basically not doable. I would otherwise say that this is necessary in order to abandon the PickleType columns.

Besides the restrictions outlined above, ORMJob and ScheduledJob behave essentially the same as before these changes. They still have a .obj property that references a Job object in python memory; it's just internally stored as a string as well, for when the ORM stops keeping track of those objects.

Most of the comments and such referencing pickling the Job column are altered or removed.

Finally, I unfortunately had to use exactly the same code in two spots (ORMJob and ScheduledJob). There might be a way to consolidate them into one spot, but that seems like something I'd need to know more about SQLAlchemy to do.

## Reviewer guidance

This still needs to run through pytest. I could not get pytest to run through all of the tests on my own machine, despite finding that the concerned tests pass individually (all of the ones that stall or break when run as a group, that is).

I don't believe this code needs any additional tests written for it, as it doesn't add functionality. I changed some old tests to work around the new expected restrictions I mentioned.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

EliKlein committed:
Initial outline for serializing jobs to JSON

EliKlein committed:
Merge branch 'develop' of https://github.com/learningequality/kolibri into develop
Updating my local code before testing

EliKlein committed:
Merge branch 'develop' of https://github.com/learningequality/kolibri into develop
in order to test with latest changes

EliKlein committed:
Main code for issue
(change pickletype to json serialization)

EliKlein committed:
Modifying tests to work with new expected code operation

EliKlein committed:
Merge branch 'develop' of https://github.com/learningequality/kolibri into develop
Getting ready to push to git and possibly make pull request

EliKlein committed:
Cleaning up old pickle stuff
Comments referencing pickle, and a line of code that's not necessary any more since ORMJobs now are in full control of when things are saved or not.

EliKlein commented:
I suppose it has to have syntax that works with all versions of python then? I didn't know it doesn't work to use two sets of keyword arguments in python 2.

rtibbles commented:
Yes - we are still supporting Python 2, as about 40% of our user base that we get telemetry data from is still using Python 2.7, so it's still very much in demand, in spite of being end of life!

EliKlein committed:
Had syntax that only works in Python 3, oops

rtibbles commented:
> Finally, I unfortunately had to use exactly the same code in two spots (ORMJob and ScheduledJob). There might be a way to consolidate them into one spot, but that seems like something I'd need to know more about SQLAlchemy to do.

This seems fine to me - if there's a good way to consolidate it, I'll give some pointers, but not something to be overly concerned about.

EliKlein committed:
Cleanup and more python 2 friendly syntax

EliKlein committed:
Forgot an important line of code
(I was struggling to run the tests at the point when I moved the working code from ORMJob to ScheduledJob)

EliKlein commented:
I think I can make a lot of these pass real quick, should I update the pull request now or wait for everything to settle?

rtibbles commented:
I probably won't review fully until Monday, so feel free to make updates!

EliKlein committed:
Final linting clean up?

EliKlein committed:
Fixed issue with ping job
Pings needed datetime as an argument, so I decided to use a different encoder when encoding the json in the first place.
Turns out pings encode json themselves, so I decided to use that same one.
Just have to make sure any future functions know that their datetime/etc. arguments will turn into strings if they're gonna be called by jobs.

EliKlein committed:
*actually* fixed issue from last commit.
I finally found a way to test things with pytest properly on my end

EliKlein commented:
I'm getting a test failure (test_inactive_kolibri in test_dbbackup.py) with my most recent commit and I'm having trouble understanding the error. Everything else is passing.

rtibbles commented on kolibri/core/tasks/exceptions.py:
```diff
@@ -15,3 +15,14 @@ class JobNotFound(Exception):
+        if message is None or message == "":
+            message = "This job had an error before it was saved to the database."
+        self.prior_type = prior_type
+        super().__init__(message)
```
For a Python 2 compatible super invocation, this needs to be called like this:
```suggestion
        super(ErrorSavedWithJob, self).__init__(message)
```

rtibbles commented on kolibri/core/tasks/test/taskrunner/test_job_running.py:
```diff
@@ -115,9 +114,10 @@ class EventProxy(object):
-    def __init__(self, *args, **kwargs):
-        self.event_id = uuid.uuid4().hex
-        EVENT_PROXY_MAPPINGS[self.event_id] = Event(*args, **kwargs)
+    def __init__(self, event_id=uuid.uuid4().hex, *args, **kwargs):
```
Should set `event_id=None` here - as the default is evaluated in module scope, not at each invocation of the `__init__` function.

Can then, in the body of the `__init__` function do this:
```
if event_id is None:
    event_id = uuid.uuid4().hex
```

EliKlein committed:
Python 2 compatibility

EliKlein commented:
That should do it then!

vkWeb commented on kolibri/core/analytics/utils.py:
```diff
@@ -423,6 +423,9 @@ def perform_ping(started, server=DEFAULT_SERVER_URL):
 
     language = get_device_setting("language_id", "")
 
+    if not isinstance(started, datetime.datetime):
+        started = datetime.datetime.strptime(started)
+
```
We'll need to supply a second argument that tells the format of the date string to `strptime` like `strptime(started, "%d %m %y")` to work. 

It would be better to use dateutil's parser. Their parser tries best to figure out the format itself. 
```
from dateutil.parser import parse
started = parse(started)
```

rtibbles replied:
I think the issue here is that the JSON deserialization is not properly coercing objects back to their type - as we want to allow datetime objects to be serialized to JSON, we should be properly deserializing them from JSON too, and not passing them as string objects to functions that are expecting datetime objects.

We should be using this function to do the parsing: https://docs.djangoproject.com/en/1.11/ref/utils/#django.utils.dateparse.parse_datetime as that directly correlates with the DjangoJSONEncoder that is being used to dump the JSON in the first place.

We should be systematically deserializing the JSON to Python objects though, so that these implementation details are opaque to functions using this.

vkWeb replied:
@rtibbles :100:.

EliKlein replied:
If all of the serialization/deserialization needs to retain typings for things that can't be serialized (so far just dates and errors), does this mean we should have some sort of type-decoding dictionary, and an extra, reserved attribute in the serialized table that lists the type of each part that isn't a primitive?

I saw people warning against checking each value to see if it's a datetime because that's slow and can, in rare cases, yield false positives. Following that advice, there would be only two ways to do it; first, checking in each function that is expecting a datetime (or error); second, storing more data in the object, and using that data to decode any objects that were turned into just strings.

I'm kinda liking the idea of going back and doing the second one, it's just a bit harder initially.

rtibbles replied:
This is one reason why moving more of the Job data into separate ORMJob columns might be advantageous, as it would reduce the overhead for this kind of type checking.

If we had a top level `exception` column on the ORMJob table (that is nullable) then we don't need a general purpose parser to handle serialized `Exception` objects.

I think the main remaining issue then is that we use a `datetime` object as a function argument and we are expecting to be able to serialize/deserialize that. I think we could use a try/except to handle deserialization, using the Django utils parse datetime.

The Python JSONDecoder object allows passing `object_hooks` which try to do further deserialization, so, for example, we would try to coerce any strings to datetimes using the Django parse function, and just return the original string if it failed.

See the documentation here: https://docs.python.org/3/library/json.html#json.JSONDecoder and an example here: https://stackoverflow.com/a/10734224 (note they are using a different parsing function).

vkWeb replied:
@rtibbles I have a different opinion on this one.

I believe the task function should take care of parsing the datetime string. The reason being we can have in rare cases some false positives. I would say, we must establish that the arguments passed to the task function will all be valid-JSON types ONLY. So, then, it will be the job of the task function to take care of parsing the datetime string. 

This will be more performant and the task function will have 100% clarity on what's going on.

rtibbles replied:
In that case, we should not use `DjangoJSONEncoder` and just use regular `json.dumps`, so as not to have any ambiguity. That means that the function initializing the pingback task would have to pre-serialize the argument.

I am fine with this if we want to be strict.

vkWeb replied:
Yes `json.dumps` should be the way to go! :rocket:

vkWeb commented on kolibri/core/tasks/exceptions.py:
```diff
@@ -15,3 +15,14 @@ class JobNotFound(Exception):
 
 class JobNotRestartable(Exception):
     pass
+
+
+class ErrorSavedWithJob(Exception):
+    def __init__(self, message, prior_type):
+        if message is None or message == "":
+            message = "This job had an error before it was saved to the database."
+        self.prior_type = prior_type
+        super(ErrorSavedWithJob, self).__init__(message)
+
+    def __str__(self):
+        return "{} Exception type: {}".format(self.message, self.prior_type)
```
I don't think we should be going ahead with a separate exception class to handle exception objects.

My recommendation would be:

Our tasks backend will save the exception's name and the message/traceback as string. Assertions for defensive programming and unit testing will then be normal string comparison, like: 
```
Kolibri plugins will have code like:

if job.exception == "TypeError"
    logger.debug("something wrong")  

Unit tests will have code like:
assert job.exception == "TypeError"
```

rtibbles replied:
I think even simpler would be just to add a rule for serialization exceptions to JSON.

It would could store the import path for the Exception class, and any relevant details (note the traceback is important here). Then the job can reserialize and create the Exception class from the details in the JSON object.

vkWeb commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +172,43 @@ def __getstate__(self):
+            if self.exception.args[0] and isinstance(self.exception.args[0], str):
+                working_dictionary["exception_message"] = self.exception.args[0]
+
+        if self.state in [State.CANCELING, State.RUNNING]:
+            logger.debug(
+                "A task is currently in action. This is a bad time to save it to the database."
+            )
```
This is not true. There's no bad time to update the job database. So, we should remove this.

vkWeb commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +172,43 @@ def __getstate__(self):
-            "result",
         ]
-        return {key: self.__dict__[key] for key in keys if key in self.__dict__}
+        working_dictionary = {key: self.__dict__[key] for key in keys if key in self.__dict__}
+
+        if self.exception is not None:
+            working_dictionary["exception_type"] = type(self.exception).__name__
+            if self.exception.args[0] and isinstance(self.exception.args[0], str):
+                working_dictionary["exception_message"] = self.exception.args[0]
+
```
Following from my above recommendation:

We should be setting `self.exception` to `self.exception.__name__`.

```
if self.exception is not None and isinstance(self.exception, Exception):
    self.exception = type(self.exception).__name__
```

vkWeb commented on kolibri/core/tasks/job.py:
```diff
@@ -185,15 +235,22 @@ def __init__(self, func, *args, **kwargs):
         if job_id is None:
             job_id = uuid.uuid4().hex
 
+        exception_type = kwargs.pop("exception_type", None)
+        if exception_type is None:
+            exception = None
+        else:
+            exception_message = kwargs.pop("exception_message", None)
+            exception = ErrorSavedWithJob(exception_message, exception_type)
+
```
Following from the above recommendation regarding not having a separate exception class: 

These all should be removed.

vkWeb commented on kolibri/core/tasks/job.py:
```diff
@@ -185,15 +235,22 @@ def __init__(self, func, *args, **kwargs):
-        self.traceback = ""
-        self.exception = None
+        self.traceback = kwargs.pop("traceback", "")
+        self.exception = exception
```
`self.exception = kwargs.pop("exception", "")`.

vkWeb commented on kolibri/core/tasks/test/taskrunner/test_worker.py:
```diff
@@ -67,7 +68,9 @@ def test_can_handle_unicode_exceptions(self, worker):
         returned_job = worker.storage.get_job(job_id)
         assert returned_job.state == "FAILED"
-        assert isinstance(returned_job.exception, TypeError)
+        assert isinstance(returned_job.exception, (TypeError, ErrorSavedWithJob))
+        if isinstance(returned_job.exception, ErrorSavedWithJob):
+            assert returned_job.exception.prior_type == "TypeError"
         assert returned_job.exception.args[0] == error_text
```
`assert returned_job.exception == "TypeError"`.

vkWeb reviewed (commented):
@EliKlein good job on this PR!  

I've written some thoughts on simplifying the approach. 
<hr>

The code that has been duplicated in two places will be best solved by having job `obj` in just one table `jobs` and the scheduling information on another table `scheduled_time`. A column on `jobs` table will foreign key into `scheduled_time` table if the job is SCHEDULED.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +172,43 @@ def __getstate__(self):
+
+    @classmethod
+    def from_json(cls, json_string):
+        working_dictionary = json.loads(json_string)
```
This is where we need to be doing more - we are using the DjangoJSONEncoder above to serialize some Python objects that otherwise will not be serialized by `json.dumps` - we should be doing the inverse here and properly deserializing everything.

Systematically extending `DjangoJSONEncoder` and an equivalent Python `JSONDecoder` object to handle these cases would allow this to be handled very systematically, and for most of the logic of this PR to be focused on this serialization and deserialization step - rather than having to reach into so many other parts of the code.

rtibbles reviewed (commented):
I think a lot of the complexity here could be reduced by focusing on consolidating serialization and deserialization logic.

I think the hybrid property could be removed, and we could just focus on having helper functions that the storage class uses for deserializing and serializing the job objects. For simplicty, that could just be methods on the job object, and we are sure to call those before setting data.

EliKlein commented:
If I'm going to save exception types as module strings, I'd like to add `stringify_type` next to `stringify_func` and change `import_stringified_func` to just `import_stringified` since it's already not checking that the returned object is callable or anything.

Does that sound good? That was the end decision, for saving exceptions, right?

rtibbles commented:
> If I'm going to save exception types as module strings, I'd like to add stringify_type next to stringify_func and change import_stringified_func to just import_stringified since it's already not checking that the returned object is callable or anything.

Yes - this seems good. I've also added a comment that perhaps we can just move the exception to a top level column on the ORMJob table, so that we don't have to handle it generically when we are parsing - although if it's handled internally to the Job object, it may matter less.

vkWeb commented:
@rtibbles any specific reasons for creating exception objects out of their stringified form? I believe exception serves just one purpose: letting the client know what's been wrong. 

So, exception type as string and its full traceback as string should be what we all need...?

rtibbles commented:
> So, exception type as string and its full traceback as string should be what we all need...?

This could well be sufficient yes - I don't think we ever reraise the exception anywhere, so being able to recreate the exception may be too much.

Even just the traceback might be sufficient.

vkWeb commented:
@rtibbles yes we don't reraise it anywhere. So let us just keep it as string.

EliKlein commented:
One of the tests in `test_worker.py` ensures that errors are allowed to have unicode characters in their data. The test checks the message of the error (arg 0) to see that it is what's expected.

I've changed the test to just check that the expected message is within the traceback (since it will be at the end); but I thought I'd mention that it is (or was) "a" usage of the Exception as an object.

vkWeb commented:
@EliKlein yes - we do check whether the exception message can contain unicode chars or not but that doesn't feel to me like a use of exception as an object. 

The real use of exception as an object will be when we want to re-raise the exception after retrieving it from the database. Currently, we don't have such a use case so we are safe to serialize exception objects into strings.

EliKlein committed:
Cursory runthrough of changes
Remove new error type
Errors saved as strings
Datetime serialized and deserialized differently; related: using normal json encoder instead of django encoder
Removed .obj property of ORMJob and ScheduledJob

EliKlein committed:
Ping time structure touchup
Tests weren't passing, so I mucked around a bit until I found a satisfactory way to have things pass.

EliKlein committed:
dateutil instead of manually writing iso format

EliKlein commented:
OK, This should be nearly done if not done.

EliKlein committed:
Merge branch 'develop' of https://github.com/learningequality/kolibri into develop

vkWeb reviewed (approved):
@EliKlein Awesome work! :tada: 

There are just some linting issues that's failing the CI, can you please fix those? here's our docs on it: https://kolibri-dev.readthedocs.io/en/develop/getting_started.html#linting-and-auto-formatting.

@rtibbles apart from linting issues, this is LGTM from my side!

EliKlein commented:
Yarn doesn't seem to recognize the fmt-backend command for me. I see what it says it wants changed, I guess I can just do that?

EliKlein committed:
linting fix

rtibbles commented:
> Yarn doesn't seem to recognize the fmt-backend command for me. I see what it says it wants changed, I guess I can just do that?

Thanks - apologies, it seems our developer docs are out of date in this regard.

```
pre-commit run --all-files
```
is the easiest invocation.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -142,12 +143,26 @@ class Job(object):
+        Creates and returns a JSON-serialized string representing this Job.
+        This is how Job objects are persisted through restarts.
+
+        Note: Uses DjangoJSONEncoder to encode things. When using a type that's
```
```suggestion
```

rtibbles replied:
These comments are no longer accurate, as we have dropped use of the DjangoJSONEncoder.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -142,12 +143,26 @@ class Job(object):
+        This is how Job objects are persisted through restarts.
+
+        Note: Uses DjangoJSONEncoder to encode things. When using a type that's
+        covered by it, the job's .func method needs to deal with that.
```
```suggestion
```

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -142,12 +143,26 @@ class Job(object):
+
+        Note: Uses DjangoJSONEncoder to encode things. When using a type that's
+        covered by it, the job's .func method needs to deal with that.
+
```
```suggestion
```

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +171,40 @@ def __getstate__(self):
             "args",
             "kwargs",
             "func",
-            "result",
```
We should save the result - it just should have to be JSON serializable too.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +171,40 @@ def __getstate__(self):
-        return {key: self.__dict__[key] for key in keys if key in self.__dict__}
+
+        if isinstance(self.exception, Exception):
+            self.exception = type(self.exception).__name__
```
We shouldn't coerce the exception here - this is where the exception gets recorded into the ORM, so we should do the coersion there:

https://github.com/learningequality/kolibri/blob/6903afec0ca3ebdf65fbf0eb96ef5348ed9fc2d2/kolibri/core/tasks/storage.py#L243

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +171,40 @@ def __getstate__(self):
+        if isinstance(self.exception, Exception):
+            self.exception = type(self.exception).__name__
+
+        if self.result is not None:
```
See above.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -156,9 +171,40 @@ def __getstate__(self):
+            string_result = json.dumps(working_dictionary)
+        except TypeError as e:
+            # One or more of the keys is not JSON-serializable.
+            logger.debug("Job '{}' is not be saved: {}".format(self.job_id, str(e)))
```
I don't think we should fail silently here - we should raise an exception here to flag that this is not JSON serializable, we should not just carry on.

rtibbles commented on kolibri/core/tasks/job.py:
```diff
@@ -142,12 +143,26 @@ class Job(object):
+        covered by it, the job's .func method needs to deal with that.
+
+        Information not saved:
+        - Results of the function, if it finished
```
This comment segment will need to be updated too.

rtibbles commented on kolibri/core/tasks/scheduler.py:
```diff
@@ -172,16 +171,16 @@ def schedule(self, dt, func, interval=0, repeat=0, *args, **kwargs):
                 interval=interval,
                 repeat=repeat,
                 scheduled_time=naive_utc_datetime(dt),
-                obj=job,
```
Can just set `saved_job=job.to_json()` in here.

rtibbles reviewed (changes requested):
Blocking:
* Save `result`
* Don't fail silently, if a Job object doesn't save it's a big issue

Good to have:
* Don't coerce exception to string inside the Job object
* Update code comments.

rtibbles commented:
Thanks for all the work here @EliKlein - there's just a couple of things that I think are a blocker to merge, but this is very nearly there!

EliKlein committed:
More touching up serialization

EliKlein committed:
Job result introduced a bug
Hadn't been popping result from job instantiation
Also did some comment rewriting

EliKlein commented:
I'm not running all of the tests here, but I ran a lot of them, including the most relevant sets, and they passed, so I think they'll probably all pass.

rtibbles commented:
Thank you @EliKlein - great work!

rtibbles commented:
If you would like to be added to our AUTHORS.md, let me know, I can update it (or feel free to make a follow up PR to do so!) https://github.com/learningequality/kolibri/blob/release-v0.15.x/AUTHORS.md
